utils/udp_client.go

```diff
@@ -35,41 +37,95 @@ type AgentClientUDP struct {
 	agent.Agent
 	io.Closer
 
-	connUDP       *net.UDPConn
+	connUDP       udpConn
 	client        *agent.AgentClient
 	maxPacketSize int                   // max size of datagram in bytes
 	thriftBuffer  *thrift.TMemoryBuffer // buffer used to calculate byte size of a span
 }
 
-// NewAgentClientUDP creates a client that sends spans to Jaeger Agent over UDP.
-func NewAgentClientUDP(hostPort string, maxPacketSize int) (*AgentClientUDP, error) {
-	if maxPacketSize == 0 {
-		maxPacketSize = UDPPacketMaxLength
+type udpConn interface {
+	Write([]byte) (int, error)
+	SetWriteBuffer(int) error
+	Close() error
+}
+
+// AgentClientUDPParams allows specifying options for initializing an AgentClientUDP. An instance of this struct should
+// be passed to NewAgentClientUDPWithParams.
+type AgentClientUDPParams struct {
+	HostPort      string
+	MaxPacketSize int
+	Logger        log.Logger
+	ResolveFunc   resolveFunc
+	DialFunc      dialFunc
```

Comment:
I think the func params should be made private, they are only needed for testing.

Reply:
Ideally, we wouldn't even have them here, because most reconnection tests are in `utils/resolved_udp_conn_test.go`, where the dialer/resolver are (private) required parameters.